### PR TITLE
feat: parameter type selector

### PR DESCRIPTION
**Pull request by simonlovesyou**
This PR adds support for parsing function parameters and exposing them in the JSON schema as `Parameter<typeof func>` ([issue 733](https://github.com/vega/ts-json-schema-generator/issues/733)). Please let me know if you see any problems with this naming, am open for suggestions.

I decided to represent the parameters as an object with the argument names as the property keys. You could argue it should be represented as an array instead but then you would lose the argument name (unless you use the JSON schema `title` property to represent the argument name, but not sure if that makes sense).

I tried to shoehorn my changes into the documentation, but not sure if I'm sure if it makes sense without making bigger documentation changes, I'd be happy to hear your thoughts on it!

With this PR merged we could potentially add support for the `default` property to the outputted JSON schema, but I think that would require bigger changes so I decided not to in this PR. We could also decide to parse the `ReturnType` of the function ([added as a todo in the src/NodeParser/FunctionParser.ts](https://github.com/vega/ts-json-schema-generator/pull/735/files#diff-0f70a1b1c7bbe968d16fc8bddb3dd508adebaa260b1cea2071e9c01f413bf485R11))

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ npm install --save ts-json-schema-generator
 ./node_modules/.bin/ts-json-schema-generator --path 'my/project/**/*.ts' --type 'My.Type.Selector'
 ```
 
-Note that different platforms (e.g. Windows) may different path separators so you may have to adjust the command above.
+Note that different platforms (e.g. Windows) may use different path separators so you may have to adjust the command above.
```
Comment (simonlovesyou) [inline README.md:26]: Bonus fix, please let me know if you'd like this as a separate PR

File: README.md [hunk 2]
```diff
@@ -171,8 +171,8 @@ fs.writeFile(output_path, schemaString, (err) => {
 -p, --path 'index.ts'
     The path to the TypeScript source file. If this is not provided, the type will be searched in the project specified in the `.tsconfig`.
 
--t, --type 'My.Type.Full.Name'
-    The type the generated schema will represent. If omitted, the generated schema will contain all
+-t, --type 'My.Type.Selector'
+    The type the generated schema will represent. Can be expressed as either as the Type Reference name or as a type utility (E.g. `Parameters<typeof func>`). If omitted, the generated schema will contain all
```
Comment (domoritz) [inline README.md:175]:
```suggestion
    The type the generated schema will represent. Can be expressed as either as the Type Reference name or as a type utility (e.g. `Parameters<typeof func>`). If omitted, the generated schema will contain all
```

File: factory/parser.ts [hunk 3]
```diff
@@ -97,7 +99,8 @@ export function createParser(program: ts.Program, config: Config, augmentor?: Pa
         .addNodeParser(new NeverTypeNodeParser())
         .addNodeParser(new ObjectTypeNodeParser())
         .addNodeParser(new AsExpressionNodeParser(chainNodeParser))
-
+        .addNodeParser(new FunctionParser(chainNodeParser))
+        .addNodeParser(withJsDoc(new ParameterParser(chainNodeParser)))
```
Comment (domoritz) [inline factory/parser.ts:103]: Why do you need to add withJsDoc? We don't add it anywhere else.

File: src/SchemaGenerator.ts [hunk 4]
```diff
@@ -151,6 +151,9 @@ export class SchemaGenerator {
                     return;
                 }
                 return;
+            case ts.SyntaxKind.ArrowFunction:
```
Comment (domoritz) [inline src/SchemaGenerator.ts:154]: What about Functions, not just arrow functions?

File: src/SchemaGenerator.ts [hunk 5]
```diff
@@ -151,6 +151,9 @@ export class SchemaGenerator {
...
+                allTypes.set(`Parameters<typeof ${this.getFullName(node, typeChecker)}>`, node);
```
Comment (domoritz) [inline src/SchemaGenerator.ts:155]: I don't understand this part. Why do we add this specific type here?

File: README.md [hunk 6]
```diff
@@ -20,10 +20,10 @@ This project is made possible by a [community of contributors](https://github.co
 
 ```bash
 npm install --save ts-json-schema-generator
-./node_modules/.bin/ts-json-schema-generator --path 'my/project/**/*.ts' --type 'My.Type.Full.Name'
+./node_modules/.bin/ts-json-schema-generator --path 'my/project/**/*.ts' --type 'My.Type.Selector'
```
Comment (domoritz) [inline README.md:23]: Let's do this in a separate pull request.

File: test/config.test.ts [hunk 7]
```diff
@@ -324,6 +324,14 @@ describe("config", () => {
         })
     );
 
+    it(
+        "arrow-function-parameters",
+        assertSchema("arrow-function-parameters", {
+            type: "Parameters<typeof myFunction>",
```
Comment (domoritz) [inline test/config.test.ts:330]: Let's just test MyType and define it to be `Parameters<typeof myFunction>` in the code. Separate out the selector test into a separate test case.

File: test/config/arrow-function-parameters/main.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @minItems 1
+ */
+export type MyNonEmptyArray<T> = T[];
+/**
```
Comment (domoritz) [inline test/config/arrow-function-parameters/main.ts:5]: Let's not have too many comments here as it makes the test more complicated than necessary.

File: test/config/arrow-function-parameters/main.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * @title String field title
+ * @minLength 10
+ * @format date-time
+ * @pattern /^\d+$/
+ */
+export type NumberValue = string;
+
+export const myFunction = (
```
Comment (domoritz) [inline test/config/arrow-function-parameters/main.ts:21]: Let's simplify the function to test the essential only.

**Review by domoritz — CHANGES_REQUESTED** [review]
Thank you for the pull request. I think it would be better to separate out the selector tests and simplify the test case. Other than that, great work!

File: src/SchemaGenerator.ts (hunk 4 above)
Comment (simonlovesyou) [reply src/SchemaGenerator.ts:154]: You're right, I missed this one. Will fix!

File: factory/parser.ts (hunk 3 above)
Comment (simonlovesyou) [reply factory/parser.ts:103]: As far as I can tell that's what I had to do for it to pick up the annotated parameters in the outputted schema. Is there any other way to do it?

File: README.md (hunk 6 above)
Comment (simonlovesyou) [reply README.md:23]: This change specifically or all the changes I've made to the README?

File: factory/parser.ts (hunk 3 above)
Comment (domoritz) [reply factory/parser.ts:103]: None of the other parsers add it so I think so. Please check.

File: README.md (hunk 6 above)
Comment (domoritz) [reply README.md:23]: Just the changes related to supporting selectors in addition to types.

File: src/SchemaGenerator.ts (hunk 5 above)
Comment (simonlovesyou) [reply src/SchemaGenerator.ts:155]: Otherwise it won't be able to find the `Parameters<typeof myFunc>` selector. If I run it without it the added test case will fail with `No root type "Parameters<typeof myFunction>" found`

File: src/SchemaGenerator.ts (hunk 4 above)
Comment (simonlovesyou) [reply src/SchemaGenerator.ts:155]: Fixed in 464023647

File: src/SchemaGenerator.ts (hunk 5 above)
Comment (domoritz) [reply src/SchemaGenerator.ts:156]: That sounds fishy. I don't have time to dig into it. Please investigate a bit to see what's going on.

Comment (simonlovesyou) [reply src/SchemaGenerator.ts:156]: It makes sense to me, the current capabilities of the lib is to find "type names" of InterfaceDeclaration, ClassDeclaration, EnumDeclaration & TypeAliasDeclaration (which the prior switch cases handle). If we're introducing something akin to a type utility (`Parameters<typeof func>`) we need to introduce that as a "name". The other option I guess would be to just use the function name instead (`func`), although I'm not super fond of it since the schema entry defines the parameters and not the function per say. WDYT?

File: test/config/arrow-function-parameters/main.ts (hunk 9 above)
Comment (simonlovesyou) [reply test/config/arrow-function-parameters/main.ts:9]: Simplified it in 7afe527

File: test/config/arrow-function-parameters/main.ts (hunk 8 above)
Comment (simonlovesyou) [reply test/config/arrow-function-parameters/main.ts]: Simplified in 82e27e2

File: README.md (hunk 6 above)
Comment (simonlovesyou) [reply README.md]: Removed the changes in b6e3c0d6, will submit a separate PR once this is merged with the documentation changes

File: test/config.test.ts (hunk 7 above)
Comment (simonlovesyou) [reply test/config.test.ts:330]: I'm not entirely sure I follow, you want me to define a type `type MyType = Parameters<typeof myFunction>` in the test fixture? I'm trying to introduce something to the lib so that the user does not have to define that type explicitly, so with the ambitions of the PR I think the current test case makes sense

File: factory/parser.ts (hunk 3 above)
Comment (simonlovesyou) [reply factory/parser.ts:103]: Looking at the list of parsers I think it actually makes sense to keep the `withJsDoc` on `ParameterParser`. Currently `TypeAliasNodeParser`, `EnumNodeParser`, `InterfaceAndClassNodeParser` & `TypeLiteralNodeParser` are wrapped with `withJsDoc`. Similar to these types of node we're not introducing support of another node (`Parameter`) that can be _also_ be annotated with jsdoc similar to the previously mentioned nodes

I've tried to show what I mean in this [ts-ast-viewer snippet](https://ts-ast-viewer.com/#code/PQKhAIAEBMFMGcDGAnAlgB1QewHbgCoCe6s4AggDaoCG84AIrIhdctQC7Z5xJrqe5wIYACh2xUgCVYAMyIlwAXnAAjLFgqxqOESMS547cAFtCAMQCuORALzKAFCPDPwoCDAQoMt8AEkcVDikPF78XEKiLuCGyABc0exoOADmADROLsiysFnWsNDysPHSchIiAJRKAHzgAEQA7gAWHLAAbjm1ukA), if you click between the two jsdoc comments you'll see it highlight "TypeAliasDeclaration" & "Parameter" in the tree view

Comment (domoritz) [reply factory/parser.ts:103]: Ahh great. Makes sense. https://ts-ast-viewer.com/ is pretty nice. I've been using https://astexplorer.net/ so far.

File: src/SchemaGenerator.ts (hunk 5 above)
Comment (domoritz) [reply src/SchemaGenerator.ts:156]: Maybe we could just not support functions and only support params and return values

```ts
function foo(a: number, b: string) {
    return a;
}

// support
export type MyType = Parameters<typeof foo>;
export type MyType2 = ReturnType<typeof foo>;

// not support
export type MyType3 = typeof foo;
```

I haven't checked how easy that would be since we do need a function type node somewhere. WHat do you think?

File: test/config.test.ts (hunk 7 above)
Comment (domoritz) [reply test/config.test.ts:330]: The way I see it (and I may be missing something), this pull request introduces two changes. First, it adds support for Parameters (and maybe ReturnType). Second, it adds support for specifying more than a type name to the cli or the `type` param. I see those as separate features that should be in two prs. I hope it doesn't make too much trouble. 

If I miss something, I'm happy to have a single pull request instead.

File: src/SchemaGenerator.ts (hunk 5 above)
Comment (simonlovesyou) [reply src/SchemaGenerator.ts:156]: > Maybe we could just not support functions and only support params and return values

That would be possible, but let me try to explain my use case & why I want this feature in the lib. I am trying to use `ts-json-schema-generator` _programmatically_ to generate JSON schema input for ✨ some ✨ function (there's more to it, but this is how I plan to use your lib). If `ts-json-schema-generator` will only support type names then I have to impose the same restrictions on the users of my lib as well where all function parameters have to be type references. For the most simple functions it will get quite silly IMO:
```
type Num = number

function add(numberA: Num, numberB: Num) { // vs. having it inline `function add(numberA: number, numberB: number)`
  return numberA + numberB
}
```
I could, in theory, construct my own source file that adds arbitrary type names for the function parameters before parsing it with `ts-json-schema-generator`, but that feels quite hacky..

I hope you see why this would be a useful part of your lib and why it makes sense, if not I'd be happy to share more information

File: test/config.test.ts (hunk 7 above)
Comment (simonlovesyou) [reply test/config.test.ts:330]: This only adds one change, and that  parsing of the function parameters and including that in the outputted JSON schema. Semantically you could say that we're introducing two changes since now it will also support `Parameters<typeof myFunction>` as a type name, but I don't think I can split this out as it currently stands into multiple PRs. 

In regards to adding support of `Parameters` & `ReturnType` as type utility functions I'd be happy to add support for it in other separate PR's (after we've resolved this PR :) )

Comment (domoritz) [reply test/config.test.ts:330]: Sounds good. Let's do that.

**Review by domoritz — APPROVED** [review]

**Comment by domoritz** [comment]
Thank you for your contribution and helping me understand it better!

**Comment by pavel-yasiukevich** [comment]
Hello @domoritz and @simonlovesyou

I don't like name "Parameters<typeof func>" since result of this expression contadicts the result of same expression in TS (object vs array). I suggest to change "Parameters" to "NamedParameters" or something like that. This will allow us to use complex notation "Parameters<typeof func>[0]" corretly in both cases.
